src/main/java/com/puppycrawl/tools/checkstyle/checks/imports/PkgControl.java

```diff
@@ -40,28 +47,130 @@
     private final PkgControl parent;
     /** The full package name for the node. */
     private final String fullPackage;
+    /**
+     * The regex pattern for partial match (exact and for subpackages) - only not
+     * null if regex is true.
+     */
+    private final Pattern patternForPartialMatch;
+    /** The regex pattern for exact matches - only not null if regex is true. */
+    private final Pattern patternForExactMatch;
+    /** If this package represents a regular expression. */
+    private final boolean regex;
 
     /**
      * Construct a root node.
      * @param pkgName the name of the package.
+     * @param regex flags interpretation of pkgName as regex pattern.
      */
-    PkgControl(final String pkgName) {
+    PkgControl(final String pkgName, final boolean regex) {
         parent = null;
+        this.regex = regex;
         fullPackage = pkgName;
+        if (regex) {
+            patternForPartialMatch = createPatternForPartialMatch();
+            patternForExactMatch = createPatternForExactMatch();
+        }
+        else {
+            patternForPartialMatch = null;
+            patternForExactMatch = null;
+        }
     }
 
     /**
-     * Construct a child node.
+     * Construct a child node. The concatenation of regular expressions needs special care:
+     * see {@link #ensureSelfContainedRegex(String, boolean)} for more details.
      * @param parent the parent node.
      * @param subPkg the sub package name.
+     * @param regex flags interpretation of subPkg as regex pattern.
      */
-    PkgControl(final PkgControl parent, final String subPkg) {
+    PkgControl(final PkgControl parent, final String subPkg, final boolean regex) {
         this.parent = parent;
-        fullPackage = parent.fullPackage + "." + subPkg;
+        if (regex || parent.regex) {
```

Comment:
It was [one of your first requests](https://github.com/checkstyle/checkstyle/pull/3438#issuecomment-250331108) that I should introduce a flag field that stores if a PkgControl keeps a regular expression. I did that and it is set like this:

        if (regex || parent.regex) {
            ...
            this.regex = true;
        }

Due to the recursive nature of PkgControl `checks` will have regex=true when it comes to `import`. The recursiveness of PkgControl is natural (and not invented by me).